src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,356 @@
+        lua_pushlstring(vm, cpath, cpath_len);
+        lua_setfield(vm, -2, "cpath");
+        lua_pop(L, 1);
+
+        // remove the "package" table
+        lua_pop(L, 1);
+        lua_pop(vm, 1);
+
+        ngx_thread_mutex_create(&ctx->mutex, ngx_cycle->log);
+    } else {
+        ctx = ctxpool->next;
+        ctxpool->next = ctxpool->next->next;
+    }
+    return ctx;
+}
+
+static void
+put_task_ctx(task_ctx_t* ctx)
+{
+    ctx->next = ctxpool->next;
+    ctxpool->next = ctx;
+    /*  clear Lua stack */
+    lua_settop(ctx->vm, 0);
+}
+
+static int
+lauxh_xcopy(lua_State *from, lua_State *to, int idx, const int allow_nil)
+{
+    switch(lua_type(from, idx))
+    {
+        case LUA_TBOOLEAN:
+            lua_pushboolean(to, lua_toboolean(from, idx));
+            return LUA_TBOOLEAN;
+
+        case LUA_TLIGHTUSERDATA:
+            lua_pushlightuserdata(to, lua_touserdata(from, idx));
+            return LUA_TLIGHTUSERDATA;
+
+        case LUA_TNUMBER:
+            lua_pushnumber(to, lua_tonumber(from, idx));
+            return LUA_TNUMBER;
+
+        case LUA_TSTRING: {
+            size_t len = 0;
+            const char *str = lua_tolstring(from, idx, &len);
+            lua_pushlstring(to, str, len);
+            return LUA_TSTRING;
+        }
+
+        case LUA_TTABLE:
+            lua_newtable(to);
+            // to positive number
+            if(idx < 0) {
+                idx = lua_gettop(from) + idx + 1;
+            }
+            lua_pushnil(from);
+            while(lua_next(from, idx))
+            {
+                if(lauxh_xcopy(from, to, -2, 0) != LUA_TNONE)
+                {
+                    if(lauxh_xcopy(from, to, -1, 0) != LUA_TNONE) {
+                        lua_rawset(to, -3);
+                    }
+                    else {
+                        lua_pop(to, 1);
+                    }
+                }
+                lua_pop(from, 1);
+            }
+            return LUA_TTABLE;
+
+        case LUA_TNIL:
+            if(allow_nil) {
+                lua_pushnil(to);
+                return LUA_TNIL;
+            }
+            // fall through
+
+        // ignore unsupported values
+        // LUA_TNONE
+        // LUA_TFUNCTION
+        // LUA_TUSERDATA
+        // LUA_TTHREAD
+        default:
+            return LUA_TNONE;
+    }
+}
+
+/* executed in a separate thread */
+static void
+my_thread_func(void *data, ngx_log_t *log)
+{
+    my_thread_ctx_t *ctx = data;
+    lua_State *vm = ctx->ctx->vm;
+
+    ctx->rc = lua_pcall(vm, ctx->n_args, LUA_MULTRET, 0);
+}
+
+/* executed in nginx event loop */
+static void
+my_thread_completion(ngx_event_t *ev)
+{
+    my_thread_ctx_t *myctx = ev->data;
+    ngx_thread_mutex_lock(&myctx->ctx->mutex, ngx_cycle->log);
+    int is_abort = myctx->is_abort;
+    ngx_thread_mutex_unlock(&myctx->ctx->mutex, ngx_cycle->log);
+    if (is_abort) {
+        put_task_ctx(myctx->ctx);
+        return;
+    }
+
+    lua_State* L = myctx->wait_co_ctx->co;

```

Comment:
Don't declare C variables in the middle of the code blocks. It would make C compilers using older C standards unhappy and it's also against the Nginx C coding style guidelines.

Reply:
Does the local var in inner scope (le.g. if block) need to be place at the top of the function?

https://openresty.org/en/c-coding-style-guide.html#local-variables

For example,
```c
if (...) {
    int a = 0;
    // any codes consuming a
}
```